src/Query/EagerLoader/Cache.php

```diff
@@ -0,0 +1,73 @@
+<?php
+namespace Mongolid\Query\EagerLoader;
+
+use Iterator;
+use MongoDB\BSON\ObjectId;
+use Mongolid\Container\Container;
+use Mongolid\Model\ModelInterface;
+use Mongolid\Util\CacheComponentInterface;
+
+/**
+ * This class is responsible for caching all related models
+ * from a previous query in order to avoid the N+1 problem.
+ * It also has a limit of models that can be cached
+ * for performance reasons.
+ */
+class Cache
+{
+    public function cache(Iterator $models, array $eagerLoadModels = []): void
+    {
+        if ($this->shouldNotCache($models, $eagerLoadModels)) {
+            return;
+        }
+
+        /** @var CacheComponentInterface $cacheComponent */
+        $cacheComponent = Container::make(CacheComponentInterface::class);
+        $extractor = new Extractor($eagerLoadModels);
+
+        // Loops through all models returned by the previous query
+        // and cache all related ids. These ids can come from
+        // embedded models or referenced models.
+        foreach ($models as $model) {
+            // The model received here may be either an array from mongodb
+            // or a cached model as a serialized array.
+            // That's why we always will force an array to be used here.
+            // It will ensure that we are only working with model arrays.
+            $model = (array) $model;
+            $extractor->extractFrom($model);
```

Comment:
If you change this method to return only related models from this specific model, maybe you'll be able to internalize the next `foreach` into this foreach, reducing iterates and gain some performance.

Reply:
```suggestion
            $loadModel = $extractor->extractFrom($model);
            
            $model = Container::make($loadModel['model']);
            $ids = array_values($loadModel['ids']);
            $query = ['_id' => ['$in' => $ids]];

            foreach ($model->where($query) as $relatedModel) {
                $cacheKey = $this->generateCacheKey($relatedModel);
                $cacheComponent->put($cacheKey, $relatedModel, 36);
            }
```

Reply:
I can't do this, because I don't know all the models I should get.
Doing this way, every loop I will increase the number of Ids.
so
loop1 [ids => [123]]
loop2 [ids => [123, 456]
...